dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp

```diff
@@ -1198,101 +1234,90 @@ std::unordered_map<TableID, DAGStorageInterpreter::StorageWithStructureLock> DAG
 
         auto lock = table_store->lockStructureForShare(context.getCurrentQueryId());
 
-        /// Check schema version, requiring TiDB/TiSpark and TiFlash both use exactly the same schema.
-        // We have three schema versions, two in TiFlash:
-        // 1. Storage: the version that this TiFlash table (storage) was last altered.
-        // 2. Global: the version that TiFlash global schema is at.
-        // And one from TiDB/TiSpark:
-        // 3. Query: the version that TiDB/TiSpark used for this query.
-        auto storage_schema_version = table_store->getTableInfo().schema_version;
-        // Not allow storage > query in any case, one example is time travel queries.
-        if (storage_schema_version > query_schema_version)
-            throw TiFlashException(
-                fmt::format("Table {} schema version {} newer than query schema version {}", table_id, storage_schema_version, query_schema_version),
-                Errors::Table::SchemaVersionError);
-        // From now on we have storage <= query.
-        // If schema was synced, it implies that global >= query, as mentioned above we have storage <= query, we are OK to serve.
+        // check the columns in table_scan and table_store, to check whether we need to sync table schema.
+        auto [are_columns_matched, error_message] = compareColumns(table_scan, table_store->getStoreColumnDefines(), dagContext(), log);
+
+        if (are_columns_matched)
+        {
+            return std::make_tuple(table_store, lock);
+        }
+
+        // columns not match but we have synced schema, it means the schema in tiflash is newer than that in query
         if (schema_synced)
-            return {table_store, lock, storage_schema_version, true};
-        // From now on the schema was not synced.
-        // 1. storage == query, TiDB/TiSpark is using exactly the same schema that altered this table, we are just OK to serve.
-        // 2. global >= query, TiDB/TiSpark is using a schema older than TiFlash global, but as mentioned above we have storage <= query,
-        // meaning that the query schema is still newer than the time when this table was last altered, so we still OK to serve.
-        if (storage_schema_version == query_schema_version || global_schema_version >= query_schema_version)
-            return {table_store, lock, storage_schema_version, true};
-        // From now on we have global < query.
-        // Return false for outer to sync and retry.
-        return {nullptr, {}, DEFAULT_UNSPECIFIED_SCHEMA_VERSION, false};
+        {
+            throw TiFlashException(fmt::format("Table {} in keyspace {} schema is newer than query schema version, columns info in query is not matches the column info in tiflash: {}", table_id, dagContext().getKeyspaceID(), error_message), Errors::Table::SchemaVersionError);
```

Comment:
```suggestion
            throw TiFlashException(fmt::format("The schema does not match the query, details: {}", error_message), Errors::Table::SchemaVersionError);
```
The error message contains the table's information